shared/web_features_manifest_github_download.go

```diff
@@ -0,0 +1,129 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package shared
+
+import (
+	"compress/gzip"
+	"context"
+	"errors"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/google/go-github/v47/github"
+)
+
+// WebFeatureManifestRepo is where the web feature manifest is published
+const WebFeatureManifestRepo = "wpt"
+
+// webFeaturesManifestFilename is the name of the manifest file in a given release
+const webFeaturesManifestFilename = "WEB_FEATURES_MANIFEST.json.gz"
+
+// ErrNoWebFeaturesManifestFileFound when a given GitHub release does not contain a manifest file to download.
+var ErrNoWebFeaturesManifestFileFound = errors.New("web features manifest not found in release")
+
+// ErrMissingBodyDuringWebFeaturesManifestDownload when a http call to download the file contains no body.
+// This should not happen often.
+var ErrMissingBodyDuringWebFeaturesManifestDownload = errors.New("empty body when downloading web features manifest")
+
+// ErrUnableToRetrieveGitHubRelease indicates the request to retrieve the latest release failed
+var ErrUnableToRetrieveGitHubRelease = errors.New("failed to retrieve latest github release")
+
+// ErrGitHubAssetDownloadFailedToComplete indicates the download request failed for a given release asset
+var ErrGitHubAssetDownloadFailedToComplete = errors.New("request to download github asset failed")
+
+// gzipBodyTransformer extracts the g-zipped body into a raw file stream.
+type gzipBodyTransformer struct{}
+
+func (t gzipBodyTransformer) Transform(body io.Reader) (io.ReadCloser, error) {
+	return gzip.NewReader(body)
+}
+
+// ResponseBodyTransformer provides a interface to transform the incoming response body into a
+// final expected body format.
+type ResponseBodyTransformer interface {
+	Transform(io.Reader) (io.ReadCloser, error)
+}
+
+// NewGitHubWebFeaturesManifestDownloader creates a downloader which will examine
+// a GitHub release, find the manifest file, and download it.
+func NewGitHubWebFeaturesManifestDownloader(httpClient *http.Client, gitHubClient *github.Client) *GitHubWebFeaturesManifestDownloader {
+	return &GitHubWebFeaturesManifestDownloader{
+		httpClient:      httpClient,
+		gitHubClient:    gitHubClient,
+		bodyTransformer: gzipBodyTransformer{},
+	}
+}
+
+// GitHubWebFeaturesManifestDownloader is a downloader that will examine
+// a GitHub release, find the manifest file, and download it.
+// Use NewGitHubWebFeaturesManifestDownloader to create an instance.
+type GitHubWebFeaturesManifestDownloader struct {
+	httpClient      *http.Client
+	gitHubClient    *github.Client
+	bodyTransformer ResponseBodyTransformer
+}
+
+// Download attempts to download the manifest file from the latest release.
+func (d GitHubWebFeaturesManifestDownloader) Download(ctx context.Context) (io.ReadCloser, error) {
+	release, _, err := d.gitHubClient.Repositories.GetLatestRelease(
+		ctx,
+		"jcscottiii", // REPLACE WITH SourceOwner BEFORE MERGING,
```

Comment:
This will be reverted back to SourceOwner once https://github.com/web-platform-tests/wpt/pull/43828 is merged